### PR TITLE
Support TIMESTAMP(6) in PostgreSql connector

### PR DESCRIPTION
**Pull request by kokosing**
Support TIMESTAMP(6) in PostgreSql connector

### CODE REVIEW

File: presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/StandardColumnMappings.java [hunk 1]
```diff
@@ -446,12 +479,38 @@ public static long toPrestoTimestamp(LocalDateTime localDateTime)
...
+    private static Instant getInstantFromTimestamp(TimestampType timestampType, long value)
+    {
+        int precision = timestampType.getPrecision();
+        if (precision >= 0 && precision <= 3) {
+            return Instant.ofEpochMilli(value);
+        }
+        if (precision > 3 && precision <= 6) {
+            return Instant.ofEpochSecond(value / 1_000_000, value % 1_000_000 * 1000);
+        }
+        throw new IllegalArgumentException("Unsupported timestamp precision: " + precision);
```
Comment (electrum) [inline presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/StandardColumnMappings.java:513]: If this goes to the user, it should be `PrestoException(NOT_SUPPORTED)`

Can you add a test that would hit this?

File: presto-testing/src/main/java/io/prestosql/testing/TestingPrestoClient.java [hunk 2]
```diff
@@ -224,7 +230,23 @@ else if (TIME_WITH_TIME_ZONE.equals(type)) {
...
+            int year = parseInt(matcher.group("year"));
+            int month = parseInt(matcher.group("month"));
+            int day = parseInt(matcher.group("day"));
+            int hour = parseInt(matcher.group("hour"));
+            int minute = parseInt(matcher.group("minute"));
+            int second = parseInt(matcher.group("second"));
+            String fraction = matcher.group("fraction");
+            fraction += "0".repeat(9 - fraction.length());
+            int fractionValue = parseInt(fraction);
+
```
Comment (martint) [inline presto-testing/src/main/java/io/prestosql/testing/TestingPrestoClient.java:247]: This needs to round to nanoseconds if precision is > 9. See VarcharToTimestampCast.castToLongTimestamp

File: presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/StandardColumnMappings.java (hunk 1 above)
Comment (martint) [reply presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/StandardColumnMappings.java:513]: This can actually never happen if we're dealing with a "short" timestamp (i.e., precision < 6, `long` representation). So it's fine to make it an IllegalArgumentException.

**Comment by kokosing** [comment]
@martint Comments addressed. Thanks for the review.

File: presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/StandardColumnMappings.java [hunk 3]
```diff
@@ -330,7 +336,9 @@ public static ColumnMapping timeColumnMapping(ConnectorSession session)
                     TIME,
                     (resultSet, columnIndex) -> {
                         LocalTime time = resultSet.getObject(columnIndex, LocalTime.class);
-                        return toPrestoLegacyTimestamp(time.atDate(LocalDate.ofEpochDay(0)), sessionZone);
+                        // disable rounding for time
```
Comment (kokosing) [inline presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/StandardColumnMappings.java:339]: Presto stores LocaTime of 23:59:59.999, then reads LocalTime 23:59:59.999999, rounding rounds to 00:00:00. 
This is coherent with `NANOSECONDS.toMillis(time.toNanoOfDay());` of non legacy time.

Comment (martint) [reply presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/StandardColumnMappings.java:339]: Ah yes, I ran into a similar issue when implementing variable precision time type: https://github.com/prestosql/presto/pull/4381. That's why I had to add `timeColumnMappingWithTruncation` for connectors that need those semantics. FYI, we're removing legacy time in that PR. Also, some of these changes will conflict with that PR, since you're modifying the parts that relate to TIME. I would leave those alone for now, since they are not related to TIMESTAMP(6).

File: presto-postgresql/src/test/java/io/prestosql/plugin/postgresql/TestPostgreSqlTypeMapping.java [hunk 4]
```diff
@@ -148,7 +148,8 @@ protected QueryRunner createQueryRunner()
                 postgreSqlServer,
                 ImmutableMap.of(),
                 ImmutableMap.of("jdbc-types-mapped-to-varchar", "Tsrange, Inet" /* make sure that types are compared case insensitively */),
-                ImmutableList.of());
+                ImmutableList.of(),
+                1);
```
Comment (findepi) [inline presto-postgresql/src/test/java/io/prestosql/plugin/postgresql/TestPostgreSqlTypeMapping.java:152]: Using 2+ nodes here is redundant, but do we gain much by requesting 1 node here?
If not, let's keep the code simpler.

File: presto-main/src/main/java/io/prestosql/testing/MaterializedResult.java [hunk 5]
```diff
@@ -387,7 +388,12 @@ else if (prestoValue instanceof SqlTimeWithTimeZone) {
                         zone);
             }
             else if (prestoValue instanceof SqlTimestamp) {
-                convertedValue = SqlTimestamp.JSON_FORMATTER.parse(prestoValue.toString(), LocalDateTime::from);
+                SqlTimestamp timestampValue = (SqlTimestamp) prestoValue;
+                convertedValue = Timestamps.getInstant(timestampValue.getEpochMicros(), toIntExact(timestampValue.getPicosOfMicros()))
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/testing/MaterializedResult.java:392]: `Timestamps.getInstant` rounds (by necessity).
It is an appropriate behavior for some cases, but for test code (eg here), we should
be more strict. We should not attempt to use `Instant` where picos matter.

So, I propose you move that method here (or inline) and instead of `roundToNearest`
you do `LongMath.divide(picosOfMicro, PICOSECONDS_PER_NANOSECOND, UNNECESSARY)`

File: presto-main/src/main/java/io/prestosql/type/Timestamps.java [hunk 6]
```diff
@@ -220,17 +220,22 @@ else if (precision <= MAX_SHORT_PRECISION) {
...
-        Instant instant = Instant.ofEpochSecond(epochSecond, nanoFraction);
+        Instant instant = getInstant(epochMicros, picosOfMicro);
         if (session.isLegacyTimestamp()) {
             return LocalDateTime.ofInstant(instant, session.getTimeZoneKey().getZoneId());
         }
 
         return LocalDateTime.ofInstant(instant, ZoneOffset.UTC);
     }
 
+    public static Instant getInstant(long epochMicros, int picosOfMicro)
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/type/Timestamps.java:236]: Let's not add this method yet. See comment in `MaterializedResult`

File: presto-spi/src/main/java/io/prestosql/spi/type/SqlTimestamp.java [hunk 7]
```diff
@@ -136,6 +131,11 @@ public SqlTimestamp roundTo(int precision)
         return newInstanceWithRounding(precision, epochMicros, picosOfMicros, sessionTimeZoneKey);
     }
 
+    public TimestampType getPrestoType()
```
Comment (findepi) [inline presto-spi/src/main/java/io/prestosql/spi/type/SqlTimestamp.java:134]: Unused, remove

File: presto-plugin-toolkit/src/main/java/io/prestosql/plugin/base/type/Timestamps.java [hunk 8]
```diff
@@ -11,7 +11,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package io.prestosql.type;
+package io.prestosql.plugin.base.type;
```
Comment (findepi) [inline presto-plugin-toolkit/src/main/java/io/prestosql/plugin/base/type/Timestamps.java:14]: We also have `Timestamps` in the SPI, so connectors now have two `Timestamps` classes available.
How do we decide which methods go into which class?

File: presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/StandardColumnMappings.java [hunk 9]
```diff
@@ -279,7 +285,7 @@ public static ColumnMapping timeColumnMappingUsingSqlTime(ConnectorSession sessi
                     TIME,
                     (resultSet, columnIndex) -> {
                         Time time = resultSet.getTime(columnIndex);
-                        return toPrestoLegacyTimestamp(toLocalTime(time).atDate(LocalDate.ofEpochDay(0)), sessionZone);
+                        return toPrestoLegacyTimestamp(createTimestampType(3), toLocalTime(time).atDate(LocalDate.ofEpochDay(0)), sessionZone);
```
Comment (findepi) [inline presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/StandardColumnMappings.java:288]: If you want to rely on the fact `createTimestampType(3)` returns a constant, just use the (deprecated)
constant directly. If you do not want to rely on that -- capture `createTimestampType(3)` as a constant outside of the lambda.

File: presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/StandardColumnMappings.java [hunk 10]
```diff
@@ -446,12 +471,36 @@ public static long toPrestoTimestamp(LocalDateTime localDateTime)
...
+    @Deprecated
+    public static LocalDateTime fromPrestoLegacyTimestamp(TimestampType timestampType, long value, ZoneId sessionZone)
+    {
+        return getInstantFromTimestamp(timestampType, value).atZone(sessionZone).toLocalDateTime();
     }
 
     public static LocalDateTime fromPrestoTimestamp(long value)
     {
-        return Instant.ofEpochMilli(value).atZone(UTC).toLocalDateTime();
+        return fromPrestoTimestamp(createTimestampType(3), value);
```
Comment (findepi) [inline presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/StandardColumnMappings.java:488]: This is called for every row, so use the TIMESTAMP const, or require that `createTimestampType(3)` be provided by the caller. (or mark the method deprecated and we remove it soon)

File: presto-plugin-toolkit/src/main/java/io/prestosql/plugin/base/type/Timestamps.java (hunk 8 above)
Comment (findepi) [reply presto-plugin-toolkit/src/main/java/io/prestosql/plugin/base/type/Timestamps.java:14]: Also, some methods from the class are specific to the engine (parsing literals),
and we should not move these.

File: presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/StandardColumnMappings.java [hunk 11]
```diff
@@ -446,12 +471,36 @@ public static long toPrestoTimestamp(LocalDateTime localDateTime)
...
+    }
+
+    private static Instant getInstantFromTimestamp(TimestampType timestampType, long value)
+    {
+        int precision = timestampType.getPrecision();
+        checkArgument(precision <= MAX_SHORT_PRECISION && precision >= 0, "Precision is out of range");
+        if (precision <= 3) {
+            return Instant.ofEpochMilli(value);
+        }
+        return Timestamps.getInstant(value, 0);
```
Comment (findepi) [inline presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/StandardColumnMappings.java:503]: This is correct, but looking at the code line, it's not obvious.
One has to check parameter names to see long value matches 
the expected representation.

File: presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/StandardColumnMappings.java (hunk 3 above)
Comment (findepi) [reply presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/StandardColumnMappings.java:339]: Does this fix a bug that we already have?
Or, we would have?

File: presto-main/src/main/java/io/prestosql/operator/scalar/timestamptz/TimestampWithTimeZoneOperators.java [hunk 12]
```diff
@@ -42,8 +42,8 @@
 import static io.prestosql.spi.type.DateTimeEncoding.unpackZoneKey;
 import static io.prestosql.spi.type.TimestampWithTimeZoneTypes.hashLongTimestampWithTimeZone;
 import static io.prestosql.spi.type.TimestampWithTimeZoneTypes.hashShortTimestampWithTimeZone;
+import static io.prestosql.spi.type.Timestamps.roundToNearest;
```
Comment (kokosing) [inline presto-main/src/main/java/io/prestosql/operator/scalar/timestamptz/TimestampWithTimeZoneOperators.java:45]: Move SPI changes to separate commit.

**Comment by kokosing** [comment]
Rebased. @martint @findepi can you please take a loot?

**Comment by kokosing** [comment]
@martint @electrum can you please take a look?

File: presto-postgresql/src/test/java/io/prestosql/plugin/postgresql/TestPostgreSqlTypeMapping.java (hunk 4 above)
Comment (findepi) [reply presto-postgresql/src/test/java/io/prestosql/plugin/postgresql/TestPostgreSqlTypeMapping.java:153]: We are not testing engine. Rather, by using distributed runner we are testing connector properly passing information between nodes -- within column handles, etc, whatever is necessary.

This is the premise for the check here
https://github.com/prestosql/presto/blob/aa6117c38bf9b9921d36a44615e7694a6f5faea0/presto-testing/src/main/java/io/prestosql/testing/AbstractTestIntegrationSmokeTest.java#L53

While I do not see a particular reason why type mapping tests _need_ that, I would rather be on the safe side.

> they are more convenient to use locally as their start up time is shorter

How much is this?

File: presto-postgresql/src/test/java/io/prestosql/plugin/postgresql/TestPostgreSqlTypeMapping.java [hunk 13]
```diff
@@ -1058,7 +1059,7 @@ private void addArrayTimestampTestIfSupported(DataTypeTest tests, DataType<List<
         };
     }
 
-    @Test(dataProvider = "testTimestampWithTimeZoneDataProvider")
```
Comment (findepi) [inline presto-postgresql/src/test/java/io/prestosql/plugin/postgresql/TestPostgreSqlTypeMapping.java:1066]: > Rename misleading testTimestampWithTimeZoneDataProvider

It was following `<test name>DataProvider` convention.
What do you find misleading about that?


(im still ok with the removal)

**Review by findepi — COMMENTED** [review]
Up until last commit. Small comments.
Can you split out the non-disputed refctorins into separate PR
so we can merge them faster?

File: presto-postgresql/src/main/java/io/prestosql/plugin/postgresql/PostgreSqlClient.java [hunk 14]
```diff
@@ -598,12 +601,22 @@ else if (elementType.getJavaType() == Slice.class) {
...
-            Array jdbcArray = statement.getConnection().createArrayOf(baseElementJdbcTypeName, getJdbcObjectArray(session, elementType, block));
+            Array jdbcArray = statement.getConnection().createArrayOf(psqlBaseTypeName, getJdbcObjectArray(session, elementType, block));
             statement.setArray(index, jdbcArray);
         });
     }
 
+    private static String getPgArrayBaseType(String baseArrayTypeName)
+    {
+        if (baseArrayTypeName.matches("timestamp\\(\\d\\)")) {
+            // parametric timestamp is not supported in PSQL JDBC driver: https://github.com/pgjdbc/pgjdbc/issues/1864
```
Comment (findepi) [inline presto-postgresql/src/main/java/io/prestosql/plugin/postgresql/PostgreSqlClient.java:614]: "array of parametric timestamp" ?

File: presto-postgresql/src/main/java/io/prestosql/plugin/postgresql/TypeUtils.java [hunk 15]
```diff
@@ -176,8 +176,9 @@ private static Object prestoNativeToJdbcObject(ConnectorSession session, Type pr
             return new Date(UTC.getMillisKeepLocal(DateTimeZone.getDefault(), millis));
         }
 
-        if (TIMESTAMP.equals(prestoType)) {
-            return toPgTimestamp(fromPrestoTimestamp((long) prestoNative));
+        if (prestoType instanceof TimestampType) {
```
Comment (findepi) [inline presto-postgresql/src/main/java/io/prestosql/plugin/postgresql/TypeUtils.java:179]:
```suggestion
        if (prestoType instanceof TimestampType && prestoType.isShort()) {
```

File: presto-postgresql/src/test/java/io/prestosql/plugin/postgresql/TestPostgreSqlTypeMapping.java [hunk 16]
```diff
@@ -1336,7 +1360,7 @@ private Session withUnsupportedType(UnsupportedTypeHandling unsupportedTypeHandl
                 TIMESTAMP_WITH_TIME_ZONE,
                 DateTimeFormatter.ofPattern("'TIMESTAMP '''yyyy-MM-dd HH:mm:ss.SSS VV''")::format,
                 // PostgreSQL does not store zone, only the point in time
-                zonedDateTime -> zonedDateTime.withZoneSameInstant(ZoneId.of("UTC")));
+                zonedDateTime -> zonedDateTime.truncatedTo(MILLIS).withZoneSameInstant(ZoneId.of("UTC")));
```
Comment (findepi) [inline presto-postgresql/src/test/java/io/prestosql/plugin/postgresql/TestPostgreSqlTypeMapping.java:1363]: why truncate here?

and why only millis, where you're adding precision up to micro?

(if this is needed, a code comment is due)

File: presto-postgresql/src/test/java/io/prestosql/plugin/postgresql/TestPostgreSqlTypeMapping.java [hunk 17]
```diff
@@ -976,41 +977,64 @@ public void testTime(boolean insertWithPresto, ZoneId sessionZone)
...
+    }
+
     @Test(dataProvider = "testTimestampDataProvider")
     public void testArrayTimestamp(boolean insertWithPresto, ZoneId sessionZone)
     {
         DataType<List<LocalDateTime>> dataType;
         DataSetup dataSetup;
         if (insertWithPresto) {
-            dataType = arrayDataType(timestampDataType());
+            dataType = arrayDataType(timestampDataType(6));
```
Comment (findepi) [inline presto-postgresql/src/test/java/io/prestosql/plugin/postgresql/TestPostgreSqlTypeMapping.java:1033]: let's have arrary of timestamp tests for 3 and 6

File: presto-postgresql/src/test/java/io/prestosql/plugin/postgresql/TestPostgreSqlTypeMapping.java [hunk 18]
```diff
@@ -1348,7 +1372,7 @@ private Session withUnsupportedType(UnsupportedTypeHandling unsupportedTypeHandl
                 // PostgreSQL does not store zone, only the point in time
                 zonedDateTime -> DateTimeFormatter.ofPattern("'TIMESTAMP WITH TIME ZONE '''yyyy-MM-dd HH:mm:ss.SSS VV''").format(zonedDateTime.withZoneSameInstant(UTC)),
                 DateTimeFormatter.ofPattern("'TIMESTAMP '''yyyy-MM-dd HH:mm:ss.SSS VV''")::format,
-                zonedDateTime -> zonedDateTime.withZoneSameInstant(ZoneId.of("UTC")));
+                zonedDateTime -> zonedDateTime.truncatedTo(MILLIS).withZoneSameInstant(ZoneId.of("UTC")));
```
Comment (findepi) [inline presto-postgresql/src/test/java/io/prestosql/plugin/postgresql/TestPostgreSqlTypeMapping.java:1375]: why truncate here?

and why only millis, where you're adding precision up to micro?

(if this is needed, a code comment is due)

File: presto-testing/src/main/java/io/prestosql/testing/datatype/DataType.java [hunk 19]
```diff
@@ -194,6 +196,23 @@
                 identity());
     }
 
+    public static DataType<LocalDateTime> timestampDataType(int precision)
+    {
+        checkArgument(precision >= 0 && precision <= 9, "Unsupported precision: %s", precision);
```
Comment (findepi) [inline presto-testing/src/main/java/io/prestosql/testing/datatype/DataType.java:201]: Why 9 and not MAX_PRECISION (12)?
Add a comment

File: presto-postgresql/src/test/java/io/prestosql/plugin/postgresql/TestPostgreSqlTypeMapping.java (hunk 18 above)
Comment (losipiuk) [reply presto-postgresql/src/test/java/io/prestosql/plugin/postgresql/TestPostgreSqlTypeMapping.java:1375]: no longer valid

File: presto-postgresql/src/test/java/io/prestosql/plugin/postgresql/TestPostgreSqlTypeMapping.java (hunk 16 above)
Comment (losipiuk) [reply presto-postgresql/src/test/java/io/prestosql/plugin/postgresql/TestPostgreSqlTypeMapping.java:1363]: not valid

File: presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/StandardColumnMappings.java [hunk 20]
```diff
@@ -349,55 +356,81 @@ public static LongWriteFunction timeWriteFunction()
...
+        checkArgument(precision <= MAX_SHORT_PRECISION && precision >= 0, "Precision is out of range");
+        Instant instant;
+        if (precision <= 3) {
+            instant = Instant.ofEpochMilli(value);
+        }
+        else {
+            long epochMicros = value;
+            long epochSecond = floorDiv(epochMicros, MICROSECONDS_PER_SECOND);
+            int nanoFraction = floorMod(epochMicros, MICROSECONDS_PER_SECOND) * NANOSECONDS_PER_MICROSECOND;
+            instant = Instant.ofEpochSecond(epochSecond, nanoFraction);
```
[the remark this answers is not in the document]
Comment (losipiuk) [reply presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/StandardColumnMappings.java:431]: seems to work according to tests

File: presto-postgresql/src/main/java/io/prestosql/plugin/postgresql/PostgreSqlClient.java (hunk 14 above)
Comment (losipiuk) [reply presto-postgresql/src/main/java/io/prestosql/plugin/postgresql/PostgreSqlClient.java:614]: dropped method. logic moved to `TypeUtils.getArrayElementPgTypeName`

**Comment by losipiuk** [comment]
Superseded by https://github.com/prestosql/presto/pull/5124
